src/PowerPing/Ping.cs

```diff
@@ -457,85 +457,50 @@ private PingResults SendICMP(PingAttributes attrs)
                     if (Display.ShowOutput) {
                         PowerPing.Display.Error("General transmit error");
                     }
-                    Results.SaveResponseTime(-1);
-                    try { Results.Lost++; }
-                    catch (OverflowException) { Results.HasOverflowed = true; }
+                    results.SaveResponseTime(-1);
+                    try { results.Lost++; }
+                    catch (OverflowException) { results.HasOverflowed = true; }
 
                 } catch (SocketException) {
 
                     PowerPing.Display.Timeout(index);
 		            if (attrs.BeepLevel == 1) {
                         try { Console.Beep(); }
-                        catch (Exception) { Results.HasOverflowed = true; }
+                        catch (Exception) { results.HasOverflowed = true; }
                     }
-                    Results.SaveResponseTime(-1);
-                    try { Results.Lost++; }
-                    catch (OverflowException) { Results.HasOverflowed = true; }
+                    results.SaveResponseTime(-1);
+                    try { results.Lost++; }
+                    catch (OverflowException) { results.HasOverflowed = true; }
+
+                } catch (OperationCanceledException) {
+
+                    break;
 
                 } catch (Exception) {
 
                     if (Display.ShowOutput) {
                         PowerPing.Display.Error("General error occured");
                     }
-                    Results.SaveResponseTime(-1);
-                    try { Results.Lost++; }
-                    catch (OverflowException) { Results.HasOverflowed = true; }
+                    results.SaveResponseTime(-1);
+                    try { results.Lost++; }
+                    catch (OverflowException) { results.HasOverflowed = true; }
 
-                } finally {
-                    // Increment seq and wait for interval
-                    index++;
-                    
-                    // Wait for set interval before sending again
-                    cancelEvent.WaitOne(attrs.Interval);
-
-                    // Generate random interval when RandomTimings flag is set
-                    if (attrs.RandomTiming) {
-                        attrs.Interval = Helper.RandomInt(5000, 100000);
-                    }
                 }
 
-                // Stop sending if flag is set
-                if (!IsRunning) {
-                    break;
-                }
+                onResultsUpdate?.Invoke(results);
```

Comment:
Now that I understand how these events/callbacks are being used for I really like them here. Maybe just add a comment to say we are firing an callback?